### PR TITLE
chore: clean up _LRHelper [DET-3270]

### PR DESCRIPTION
**Pull request by shiyuann**
## Description

Clean up LR helper in our Pytorch support. Previously, `LRScheduler` is an user-facing object that lets users set how often an LR scheduler is stepped and `_LRHelper` is an internal object that lets the Determined step an LR scheduler. Keeping two objects is unnecessarily complicate especially in that part of the `_LRHelper` handles accepting an input of a None `LRScheduler`.

There are two different ways of calling the step() of a LR scheduler.

Method 1:
```
class MyTrial:
   def __init__(context):
       self.context = context
       self.lr_scheduler = self.context.LRScheduler(optimizer=optimizer, step_mode=MANUAL_STEP)

   def train_for_step(self, ...):
      ...
      self.lr_scheduler.step()
```

Method 2:
```
class MyTrial:
   def __init__(context):
       self.context = context
       self.lr_scheduler = self.context.LRScheduler(optimizer=optimizer, step_mode=STEP_EVERY_BATCH)

   def train_for_step(self, ...):
      ...
```
In the method 2, the scheduler step() is called for each batch.

These two methods are implemented both in the merged `LRScheduler`.

## Test Plan

Use the test suite for this. No need for additional tests.

## Commentary (optional)

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: harness/determined/pytorch/_lr_scheduler.py [hunk 1]
```diff
@@ -46,76 +49,64 @@ def __init__(self, scheduler: torch.optim.lr_scheduler._LRScheduler, step_mode:
...
-                if mod == 0 or mod < aggregation_frequency:
-                    return True
         return False
 
-    def step(self) -> None:
-        check.check_eq(
-            self._lr_scheduler.scheduler._step_count,  # type: ignore
-            self._lr_scheduler_count,
-            "You cannot call `scheduler.step()` if you have configured "
+    def _call_step(self, *args: Any, **kwargs: Any) -> None:
```
Comment (aaron276h) [inline harness/determined/pytorch/_lr_scheduler.py:83]: question: Why do we need to have this as a separate functon from `step()`?

Comment (shiyuann) [reply harness/determined/pytorch/_lr_scheduler.py:83]: @aaron276h the reason for that is `step()` is only external facing and `_call_step()` is called when it's configured to use `STEP_EVERY_BATCH` or `STEP_EVERY_EPOCH`.

Comment (shiyuann) [reply harness/determined/pytorch/_lr_scheduler.py:83]: There are two different ways of using it.
Method 1:
```
class MyTrial:
   def __init__(context):
       self.context = context
       self.lr_scheduler = self.context.LRScheduler(optimizer=optimizer, step=MANUAL_STEP)

   def train_for_step(self, ...):
      ...
      self.lr_scheduler.step()
```

Method 2:
```
class MyTrial:
   def __init__(context):
       self.context = context
       self.lr_scheduler = self.context.LRScheduler(optimizer=optimizer, step=STEP_EVERY_BATCH)

   def train_for_step(self, ...):
      ...
```
In the method 2, the scheduler step() is called for each batch.

Comment (aaron276h) [reply harness/determined/pytorch/_lr_scheduler.py:83]: We should avoid having other classes call private functions. I would just unify the logic of `_call_step()` and `step()` and perform the step check only when we control the steps.

File: harness/determined/pytorch/_lr_scheduler.py [hunk 2]
```diff
@@ -46,76 +49,64 @@ def __init__(self, scheduler: torch.optim.lr_scheduler._LRScheduler, step_mode:
...
+        """
+        _call_step is a function for internal training loop to call step() on the
+        original LRScheduler if it is not in MANUAL_STEP mode.
+        """
+        check.not_eq(
+            self.step_mode,
+            LRScheduler.StepMode.MANUAL_STEP,
+            "Please use the MANUAL_STEP step mode to call step() on the scheduler.",
+        )
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/_lr_scheduler.py:92]: It seems like this is either not a necessary check or we need to change the wording on it.

File: harness/determined/pytorch/_pytorch_trial.py [hunk 3]
```diff
@@ -431,12 +429,12 @@ def _train_for_step(self, step_id: int, batches_per_step: int) -> workload.Respo
...
                 self.context.optimizer.zero_grad()
 
-                if self.lr_helper.should_step_lr(
+                if self.context.lr_scheduler.should_step_lr(
                     batches_completed=batch_idx + 1,
                     epoch_length=len(self.training_loader),
                     aggregation_frequency=self.hvd_config.aggregation_frequency,
                 ):
-                    self.lr_helper.step()
+                    self.context.lr_scheduler.step()
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/_pytorch_trial.py:437]: Three questions:
1. should `should_step_lr()` be `_should_call_step()`?
2. should `step()` be `_call_step()`?
3. Why not just put all of this logic into `_call_step()`?

Comment (shiyuann) [reply harness/determined/pytorch/_pytorch_trial.py]: Given that the original code in this PR is confusing, I moved the code of the original `_LRHelper` to `PytorchTrial`.

**Comment by shiyuann** [comment]
@aaron276h @rb-determined-ai Given that the original code in this PR is confusing, I moved the code of the original `_LRHelper` to `PytorchTrial`.

File: harness/determined/pytorch/_pytorch_trial.py [hunk 4]
```diff
@@ -431,12 +429,20 @@ def _train_for_step(self, step_id: int, batches_per_step: int) -> workload.Respo
                     self.context.optimizer.step()
                 self.context.optimizer.zero_grad()
 
-                if self.lr_helper.should_step_lr(
-                    batches_completed=batch_idx + 1,
-                    epoch_length=len(self.training_loader),
-                    aggregation_frequency=self.hvd_config.aggregation_frequency,
-                ):
-                    self.lr_helper.step()
+                # Step learning rate of a LRScheduler.
```
[the remark this answers is not in the document]
Comment (shiyuann) [reply harness/determined/pytorch/_pytorch_trial.py:432]: I can move this part to a separate function in `PytorchTrial` if you insist. The reason I moved this part to the trial class is that the previous `_LRHelper` makes an assumption on the execution order of `PytorchTrial`, which is awkward.

[the remark this answers is not in the document]
Comment (aaron276h) [reply harness/determined/pytorch/_pytorch_trial.py:432]: Moving it to a separate function shouldn't change the order execution.

[the remark this answers is not in the document]
Comment (shiyuann) [reply harness/determined/pytorch/_pytorch_trial.py:432]: Yes. I mean if a class makes an assumption on the execution of another class the code would be hard to understand and maintain.

[the remark this answers is not in the document]
Comment (aaron276h) [reply harness/determined/pytorch/_pytorch_trial.py:432]: Functions that are 130+ LOC are also hard to maintain. I put it as non-blocking so it's your call in the end 😄 .

File: harness/determined/pytorch/_lr_scheduler.py [hunk 5]
```diff
@@ -46,76 +51,31 @@ def __init__(self, scheduler: torch.optim.lr_scheduler._LRScheduler, step_mode:
...
+            self._step_count,
+            "If you have configured Determined to manage the learning rate scheduler "
+            "step, i.e. not using MANUAL_STEP mode, you cannot call step() of the "
+            "original LR scheduler.",
         )
-        return self.scheduler.step(*args, **kwargs)
 
-    def get_last_lr(self) -> typing.List:
-        """Return last computed learning rate by current scheduler.
+        self._step_count += 1
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/_lr_scheduler.py:70]: non-blocking: I'm ok leaving this check in for now, while we have the `create_lr_scheduler()` API, but in the future state where users are calling `self.context.lr_scheduler()` or whatever we end up naming it, I'd just as soon we leave this check out entirely, since it is introspecting the sub-scheduler's private members.

File: harness/determined/pytorch/_pytorch_trial.py [hunk 6]
```diff
@@ -708,8 +715,8 @@ def _save(self, path: pathlib.Path) -> workload.Response:
             "optimizer_state_dict": self.context.optimizer.state_dict(),
         }
 
-        if self.lr_helper:
-            checkpoint["lr_scheduler"] = self.lr_helper.state_dict()
+        if self.context.lr_scheduler:
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/_pytorch_trial.py:718]: non-blocking: use `if self.context.lr_scheduler is not None`, the way that it's done in other lines of this PR.  Also I just prefer being explicit about a None check.

File: harness/determined/pytorch/_lr_scheduler.py [hunk 7]
```diff
@@ -20,15 +28,12 @@ class StepMode(enum.Enum):
...
-        Usage of this wrapper is required to properly schedule the optimizer's learning rate.
-
-        This wrapper fulfills two main functions:
-            1. Save and restore the learning rate when a trial is paused, preempted, etc.
-            2. Step the learning rate scheduler at the configured frequency
-               (e.g., every batch or every epoch).
+    def __init__(
+        self,
+        scheduler: torch.optim.lr_scheduler._LRScheduler,
+        step_mode: StepMode = StepMode.MANUAL_STEP,
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/_lr_scheduler.py:34]: blocking: we originally had a default StepMode, but users ran into issues with that case, and we made it a required parameter.  I personally think it should stay a required parameter.  Perhaps as an alternative to having a default mode, we could introduce another classes like `det.pytorch.AutoStepLRScheduler()` that users are not allowed to call `step()` on at all.

But in this PR, leaving the user-facing API unchanged seems pretty reasonable to me.
